### PR TITLE
parallel composite upload support

### PR DESCRIPTION
**Pull request by singhravidutt**
This PR adds the support for integrating with new write APIs introduced in java-storage to leverage ParallelCompositeUpload (PCU) feature.

### CODE REVIEW

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 1]
```diff
@@ -481,6 +482,49 @@ public class GoogleHadoopFileSystemConfiguration {
...
+          AsyncWriteChannelOptions.DEFAULT.getPCUBufferCount());
+
+  /**
+   * Configuration key to configure the buffer capacity for UploadType.PARALLEL_COMPOSITE_UPLOAD. It
+   * is in alignment with configuration of java-storage client
+   * https://cloud.google.com/java/docs/reference/google-cloud-storage/latest/com.google.cloud.storage.ParallelCompositeUploadBlobWriteSessionConfig.BufferAllocationStrategy#com_google_cloud_storage_ParallelCompositeUploadBlobWriteSessionConfig_BufferAllocationStrategy_fixedPool_int_int_
+   */
+  public static final HadoopConfigurationProperty<Long> GCS_PCU_BUFFER_CAPACITY =
+      new HadoopConfigurationProperty<>(
+          "fs.gs.write.parallel.composite.upload.part.file.cleanup.type",
```
Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:502]: cleanup.capactiy?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 2]
```diff
@@ -481,6 +482,49 @@ public class GoogleHadoopFileSystemConfiguration {
       GCS_WRITE_TEMPORARY_FILES_PATH =
           new HadoopConfigurationProperty<>("fs.gs.write.temporary.dirs", ImmutableSet.of());
 
+  /**
+   * Configuration key to configure the Buffers for UploadType.PARALLEL_COMPOSITE_UPLOAD. It is in
+   * alignment with configuration of java-storage client
+   * https://cloud.google.com/java/docs/reference/google-cloud-storage/latest/com.google.cloud.storage.ParallelCompositeUploadBlobWriteSessionConfig.BufferAllocationStrategy#com_google_cloud_storage_ParallelCompositeUploadBlobWriteSessionConfig_BufferAllocationStrategy_fixedPool_int_int_
+   */
+  public static final HadoopConfigurationProperty<Integer> GCS_PCU_BUFFER_COUNT =
```
Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:490]: update Configuration.md as well?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 3]
```diff
@@ -588,6 +632,10 @@ private static AsyncWriteChannelOptions getWriteChannelOptions(Configuration con
         .setUploadType(GCS_CLIENT_UPLOAD_TYPE.get(config, config::getEnum))
         .setTemporaryPaths(
             ImmutableSet.copyOf(GCS_WRITE_TEMPORARY_FILES_PATH.getStringCollection(config)))
+        .setPCUBufferCapacity(GCS_PCU_BUFFER_COUNT.get(config, config::getInt))
```
Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:635]: BufferCount?

Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:635]: Please add UTs.

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 4]
```diff
@@ -588,6 +632,10 @@ private static AsyncWriteChannelOptions getWriteChannelOptions(Configuration con
...
+        .setPCUBufferCapacity(toIntExact(GCS_PCU_BUFFER_CAPACITY.get(config, config::getLongBytes)))
+        .setPartFileCleanupType(GCS_PCU_PART_FILE_CLEANUP_TYPE.get(config, config::getEnum))
+        .setPartFilePrefix(GCS_PCU_PART_FILE_NAME_PREFIX.get(config, config::get))
```
Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:638]: rename to setPartFileNamePrefix?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 5]
```diff
@@ -481,6 +482,49 @@ public class GoogleHadoopFileSystemConfiguration {
...
+              "fs.gs.write.parallel.composite.upload.part.file.cleanup.type",
+              AsyncWriteChannelOptions.DEFAULT.getPartFileCleanupType());
+
+  /**
+   * Configuration key to set up the naming strategy of part files created via
+   * UploadType.PARALLEL_COMPOSITE_UPLOAD. It is in alignment with configuration of java-storage
+   * client
+   * https://cloud.google.com/java/docs/reference/google-cloud-storage/latest/com.google.cloud.storage.ParallelCompositeUploadBlobWriteSessionConfig.PartNamingStrategy
+   */
+  public static final HadoopConfigurationProperty<String> GCS_PCU_PART_FILE_NAME_PREFIX =
```
Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:523]: Should we validate that this does not contain invalid character for filename?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java [hunk 6]
```diff
@@ -251,12 +257,43 @@ private static BlobWriteSessionConfig getSessionConfig(AsyncWriteChannelOptions
...
+                    .setNameFormat("gcsio-storage-client-pcu-pool-%d")
+                    .build());
+        return BlobWriteSessionConfigs.parallelCompositeUpload()
+            .withBufferAllocationStrategy(
+                BufferAllocationStrategy.fixedPool(
+                    writeOptions.getPCUBufferCount(), writeOptions.getPCUBufferCapacity()))
+            .withPartCleanupStrategy(getPartCleanupStrategy(writeOptions.getPartFileCleanupType()))
+            .withExecutorSupplier(ExecutorSupplier.useExecutor(pcuThreadPool))
+            .withPartNamingStrategy(getPartNamingStrategy(writeOptions.getPartFilePrefix()));
       default:
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:273]: Is this considered an error case?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java [hunk 7]
```diff
@@ -251,12 +257,43 @@ private static BlobWriteSessionConfig getSessionConfig(AsyncWriteChannelOptions
...
     }
   }
 
+  private static PartCleanupStrategy getPartCleanupStrategy(PartFileCleanupType cleanupType) {
+    switch (cleanupType) {
+      case NEVER:
+        return PartCleanupStrategy.never();
+      case ON_SUCCESS:
+        return PartCleanupStrategy.onlyOnSuccess();
+      default:
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:285]: let us explicity add Always case and throw error in the default case?

File: gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java [hunk 8]
```diff
@@ -481,6 +482,49 @@ public class GoogleHadoopFileSystemConfiguration {
...
+          "fs.gs.write.parallel.composite.upload.part.file.cleanup.type",
+          (long) AsyncWriteChannelOptions.DEFAULT.getPCUBufferCapacity());
+
+  /**
+   * Configuration key to clean up strategy of part files created via
+   * UploadType.PARALLEL_COMPOSITE_UPLOAD. It is in alignment with configuration of java-storage
+   * client
+   * https://cloud.google.com/java/docs/reference/google-cloud-storage/latest/com.google.cloud.storage.ParallelCompositeUploadBlobWriteSessionConfig.PartCleanupStrategy
+   */
+  public static final HadoopConfigurationProperty<PartFileCleanupType>
```
Comment (arunkumarchacko) [inline gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfiguration.java:511]: why not directly use PartCleanupStrategy?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java [hunk 9]
```diff
@@ -251,12 +257,43 @@ private static BlobWriteSessionConfig getSessionConfig(AsyncWriteChannelOptions
             writeOptions.getTemporaryPaths().stream()
                 .map(x -> Paths.get(x))
                 .collect(ImmutableSet.toImmutableSet()));
+      case PARALLEL_COMPOSITE_UPLOAD:
+        ExecutorService pcuThreadPool =
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:261]: when will this be shutdown?

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java [hunk 10]
```diff
@@ -251,12 +257,43 @@ private static BlobWriteSessionConfig getSessionConfig(AsyncWriteChannelOptions
...
+            Executors.newCachedThreadPool(
+                new ThreadFactoryBuilder()
+                    .setNameFormat("gcsio-storage-client-pcu-pool-%d")
+                    .build());
+        return BlobWriteSessionConfigs.parallelCompositeUpload()
+            .withBufferAllocationStrategy(
+                BufferAllocationStrategy.fixedPool(
+                    writeOptions.getPCUBufferCount(), writeOptions.getPCUBufferCapacity()))
+            .withPartCleanupStrategy(getPartCleanupStrategy(writeOptions.getPartFileCleanupType()))
+            .withExecutorSupplier(ExecutorSupplier.useExecutor(pcuThreadPool))
```
Comment (arunkumarchacko) [inline gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:271]: this should ideally be hidden from the client? What will happen if we do not provide the threadpool? i.e. will it create  one on its own?

File: util/src/main/java/com/google/cloud/hadoop/util/AsyncWriteChannelOptions.java [hunk 11]
```diff
@@ -47,12 +54,19 @@ public enum UploadType {
     /* Write whole file to disk and then upload.*/
     WRITE_TO_DISK_THEN_UPLOAD,
     /* Write chunks to file along with uploading to gcs, and failure will be retried from data on disk.*/
-    JOURNALING
+    JOURNALING,
+    /* Write are performed using parallel composite upload strategy.  */
+    PARALLEL_COMPOSITE_UPLOAD
   }
 
+  private static final int PARALLEL_COMPOSITE_UPLOAD_BUFFER_COUNT = 1;
```
Comment (arunkumarchacko) [inline util/src/main/java/com/google/cloud/hadoop/util/AsyncWriteChannelOptions.java:62]: why 1? A comment will be helpful how we came up with this magic number.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 12]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+    gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+
+    byte[] bytesToWrite = new byte[partFileCount * bufferCapacity];
+    GoogleCloudStorageTestHelper.fillBytes(bytesToWrite);
+    WritableByteChannel writeChannel = gcs.create(resourceId);
+    writeChannel.write(ByteBuffer.wrap(bytesToWrite));
+    // part files are getting uploaded in async thread
+    // wait for it to complete before listing files
+    Thread.sleep(5000);
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:268]: This should be avoided. This will increase our test run time. Wont' close() delete this?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 13]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                pcuDefaultOptions.toBuilder()
+                    .setPartFilePrefix(partFilePrefix)
+                    .setPartFileCleanupType(PartFileCleanupType.NEVER)
+                    .build())
+            .build();
+
+    gcs = getGCSImpl(storageOptions);
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:243]: close this. Also at other places.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 14]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+      int afterCleanup,
+      String partFilePrefix)
+      throws IOException, InterruptedException {
+    byte[] bytesToWrite = new byte[beforeCleanup * bufferCapacity];
+    GoogleCloudStorageTestHelper.fillBytes(bytesToWrite);
+    WritableByteChannel writeChannel = gcs.create(resourceId);
+    writeChannel.write(ByteBuffer.wrap(bytesToWrite));
+    // part files are getting uploaded in async thread
+    // wait for it to complete before listing files
+    Thread.sleep(5000);
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:366]: This is a test smell.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 15]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+    return itemInfos.stream()
+        .filter(x -> x.getObjectName().endsWith(".part"))
+        .collect(Collectors.toList());
+  }
+
+  private void writeAndVerifyPartFiles(
+      int bufferCapacity,
+      StorageResourceId resourceId,
+      int beforeCleanup,
+      int afterCleanup,
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:357]: rename this to expectedCountAfterUpload?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 16]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+    writeChannel.close();
+    partFiles = getPartFiles(partFilePrefix);
+    // part files are deleted once upload is finished.
+    assertThat(partFiles.stream().count()).isEqualTo(afterCleanup);
+  }
+
   private void writeAndVerifyTemporaryFiles(
       StorageResourceId resourceId, int expectedTemporaryFileCount) throws IOException {
-    byte[] bytesToWrite = new byte[1024 * 1024 * 3];
+    byte[] bytesToWrite = new byte[1024 * 1024 * 200];
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:379]: why this change?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 17]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+    // wait for it to complete before listing files
+    Thread.sleep(5000);
+
+    List<GoogleCloudStorageItemInfo> partFiles = getPartFiles(partFilePrefix);
+
+    assertThat(partFiles.stream().count()).isEqualTo(beforeCleanup);
+    writeChannel.close();
+    partFiles = getPartFiles(partFilePrefix);
+    // part files are deleted once upload is finished.
+    assertThat(partFiles.stream().count()).isEqualTo(afterCleanup);
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:374]: Should we also assert that the final file exists and the content is the same?

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 18]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+
+  @Test
+  public void uploadViaPCUVerifyPartFileNotCleanedUp() throws IOException, InterruptedException {
+    String partFilePrefix = name.getMethodName();
+    GoogleCloudStorageOptions storageOptions =
+        GoogleCloudStorageTestHelper.getStandardOptionBuilder()
+            .setWriteChannelOptions(
+                pcuDefaultOptions.toBuilder()
+                    .setPartFilePrefix(partFilePrefix)
+                    .setPartFileCleanupType(PartFileCleanupType.NEVER)
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:239]: When would customer use NEVER?

Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:239]: ON_SUCCESS is not tested?

File: gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfigurationTest.java [hunk 19]
```diff
@@ -115,6 +116,11 @@ public class GoogleHadoopFileSystemConfigurationTest {
           put("fs.gs.working.dir", "/");
           put("fs.gs.client.upload.type", UploadType.DEFAULT);
           put("fs.gs.write.temporary.dirs", ImmutableSet.of());
+          put("fs.gs.write.parallel.composite.upload.buffer.count", 1);
```
Comment (arunkumarchacko) [inline gcs/src/test/java/com/google/cloud/hadoop/fs/gcs/GoogleHadoopFileSystemConfigurationTest.java:119]: test buffercapacity as well.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 20]
```diff
@@ -189,9 +213,170 @@ public void uploadViaJournaling() throws IOException {
...
+    gcs = getGCSImpl(storageOptions);
+    StorageResourceId resourceId = new StorageResourceId(TEST_BUCKET, name.getMethodName());
+
+    byte[] bytesToWrite = new byte[partFileCount * bufferCapacity];
+    GoogleCloudStorageTestHelper.fillBytes(bytesToWrite);
+    WritableByteChannel writeChannel = gcs.create(resourceId);
+    writeChannel.write(ByteBuffer.wrap(bytesToWrite));
+    // part files are getting uploaded in async thread
+    // wait for it to complete before listing files
+    Thread.sleep(5000);
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:298]: Please change this.

**Review by arunkumarchacko — CHANGES_REQUESTED** [review]

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 21]
```diff
@@ -59,10 +70,23 @@ public class GoogleCloudStorageClientImplIntegrationTest {
   private static final String GCS_WRITE_TMP_DIR_1 =
       String.format("%s/%s", TEMP_DIR_PATH, "gcs-write-dir-1");
 
+  private static final int ONE_MiB = 1024 * 1024;
+
   private static GoogleCloudStorage helperGcs;
 
   private GoogleCloudStorage gcs;
 
+  private int partFileCount = 2;
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:79]: final

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java [hunk 22]
```diff
@@ -59,10 +70,23 @@ public class GoogleCloudStorageClientImplIntegrationTest {
...
+  private int bufferCapacity = partFileCount * ONE_MiB;
```
Comment (arunkumarchacko) [inline gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:80]: final

File: util/src/main/java/com/google/cloud/hadoop/util/AsyncWriteChannelOptions.java (hunk 11 above)
Comment (singhravidutt) [reply util/src/main/java/com/google/cloud/hadoop/util/AsyncWriteChannelOptions.java:63]: added.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java (hunk 13 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java:244]: Code snippet where it is getting closed. Not sure if I understand the concern.
```
  @After
  public void cleanUp() {
    // cleanup any leaked files
    ImmutableSet<String> tempDirs = ImmutableSet.of(GCS_WRITE_TMP_DIR_1, GCS_WRITE_TMP_DIR);
    .
    .
    .

    // close cloudStorage to free up resources
    if (gcs != null) {
      gcs.close();
    }
```

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java (hunk 22 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java]: updated.

File: gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java (hunk 21 above)
Comment (singhravidutt) [reply gcsio/src/test/java/com/google/cloud/hadoop/gcsio/integration/GoogleCloudStorageClientImplIntegrationTest.java]: updated.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java (hunk 7 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java]: updated.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java (hunk 6 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java:279]: updated.

File: gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java (hunk 10 above)
Comment (singhravidutt) [reply gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java]: updated.

**Review by arunkumarchacko — APPROVED** [review]
